### PR TITLE
Raffle Machine (for #107)

### PR DESCRIPTION
**Pull request by huntertur**
For #107. The raffle machine is available under the Events drop-down.

An Array is made with the names of every SSE member. Each person is added to the Array as many times as they have total raffle points. The Spin button selects a random person from this weighted list.

This PR isn't ready for production as-is. For one, even though this is supposed to be a raffle machine, it is little more than an unstylized button and paragraph element. Someone with graphic design experience might be able to make something that actually looks like a machine :stuck_out_tongue:

### CODE REVIEW

File: _events/raffle.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+layout: page
+title: Raffle
+permalink: /events/raffle
+---
+
+<script src="/assets/js/raffle.js"></script>
+<noscript><p>JavaScript is required to use the raffle machine.</p></noscript>
+<div id="raffle"></div>
```
Comment (andy-wojciechowski) [inline _events/raffle.html]: Can you please add a loading message here before the page initially loads and then hide it when all of the points data has been loaded? See the points page for an example on how to do that.

File: assets/js/raffle.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    let spinButton = container.appendChild(document.createElement('button'));
+    spinButton.type = 'button';
+    spinButton.id = 'spin';
+    spinButton.textContent = 'Spin';
+
+    let winnerElement = container.appendChild(document.createElement('p'));
+    winnerElement.id = 'winner';
+
+    spinButton.addEventListener('click', () => {
+        winnerElement.textContent = getRandomArrayItem(entries);
```
Comment (andy-wojciechowski) [inline assets/js/raffle.js]: You should remove one of the student's entries if they win once. By the looks of this code it seems like each student can win an indefinite amount of times.

Comment (huntertur) [reply assets/js/raffle.js]: Fixed in commit f6572438621d2b8721dc169c92cc06dc98b06768.

File: _events/raffle.html (hunk 1 above)
Comment (huntertur) [reply _events/raffle.html]: Fixed in commit e21492a50546a29bdf814652e89762ccc1ffb69d.

File: assets/js/raffle.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            spinButton.disabled = true;
+            winnerElement.textContent = 'Nobody is left! Everyone has won!';
+            return;
+        }
+
+        let winner = getRandomArrayItem(entries);
+        winnerElement.textContent = winner;
+
+        /* Remove all of the winner's names from the raffle pool. */
+        entries = entries.filter(name => name !== winner);
```
Comment (andy-wojciechowski) [inline assets/js/raffle.js]: Don't remove all of a person's entries if they win once, just a single entry. People can win multiple times in the SSE end of quarter raffle and if a person wins once that means one of their entries is removed.

Comment (huntertur) [reply assets/js/raffle.js]: Ah, OK. I wasn't fully sure how the raffle system worked.

**Comment by andy-wojciechowski** [comment]
Alright, I approve of this initial raffle functionality. I'd say lets leave this open until the design on the front-end is changed. My suggestions are to display all of the students that have an entry in the raffle, and how many entries they have left.  I don't really have a design in mind nor am I good in graphic design. Some people who might be able to give you some direction on the design are Grace since she wrote this work item to begin with, or George Burkhart.

**Comment by msoe-sse-webmaster** [comment]
> Alright, I approve of this initial raffle functionality. I'd say lets leave this open until the design on the front-end is changed. My suggestions are to display all of the students that have an entry in the raffle, and how many entries they have left. I don't really have a design in mind nor am I good in graphic design. Some people who might be able to give you some direction on the design are Grace since she wrote this work item to begin with, or George Burkhart.

Okay, so I can totally try and put together some UI mocks. In terms of making it happen, I'd be happy to volunteer @rgeorgebu , but I'd also definitely recommend that you look into some UI ideas on your own :) This is a great opportunity to learn to make your frontend sparkle.

**Comment by andy-wojciechowski** [comment]
@msoe-sse-webmaster What I meant by my previous comment is I think one of you should help give @huntertur some sort of high level idea for the design and see if he can take it anywhere. Like above I gave the idea of displaying the student names and the number of entries. Maybe you can expand that but I don't think we should put together UI mockups for him to implement.

**Comment by GFLEMING133** [comment]
Yeah @andy-wojciechowski that's an important thought, and honestly I'd rather see creativity here than strict 'make it look like this' commits. So @huntertur , here are some ideas for raffle UIs:

https://wheelofnames.com/ (see below)
image

image

I think in general we want to show
1. Some sort of indication that the winner is "being chosen"
 > this could be an animation of all the names in a circle/line/oval
 > this could technically be a loading icon

2. The winner 

Does this help give you more direction? Don't feel like you need to copy either of two examples--they're just ideas.

**Comment by huntertur** [comment]
From the expanded commit message:

This is just a transition from the almost-blank webpage to something
that sort of looks like the final product. There is no spinning
animation yet, and the machine simply pops the first name off the
(randomized) list whenever the Spin button is pressed.

What comes next is probably:
 1. Make the reel spin upwards whenever the button is pressed before
    landing on a name. In other words: have animation.
 2. Randomly pick which name will be chosen instead of it being the
    first name in the list, which is already visible. Would probably
    the spinner to spin for a known amount of seconds, plus or minus a
    small random amount.

**Comment by andy-wojciechowski** [comment]
I'm closing this for now as this still needs some work, I'll put some more information in the issue about what I found we'll create a new PR once those changes have been made.

**Pull request closed without merging**